### PR TITLE
ELSI: improve package

### PR DESCRIPTION
**Pull request by RMeli**
* Add `test` and `test_c` variants to enable tests
* Add `internal_elpa_version` variant to select version of internally provided ELPA
* Made a `CudaPackage`
* Fix compilation (tested with both `openblas` and `intel-oneapi-mkl`)

### CODE REVIEW

**Comment by RMeli** [comment]
spack blame elsi

**Comment by RMeli** [comment]
@spackbot fix style

File: var/spack/repos/builtin/packages/elsi/package.py [hunk 1]
```diff
@@ -33,27 +33,51 @@ class Elsi(CMakePackage):
     variant("enable_sips", default=False, description="Enable SLEPc-SIPs support")
     variant("use_external_elpa", default=False, description="Build ELPA using SPACK")
     variant("use_external_ntpoly", default=False, description="Build NTPoly using SPACK")
-    variant("use_external_omm", default=False, description="Use external libOMM and MatrixSwitch")
     variant("use_external_superlu", default=False, description="Use external SuperLU DIST")
     variant(
         "use_mpi_iallgather", default=True, description="Use non-blocking collective MPI functions"
     )
+    variant("tests", default=False, description="Enable Fortran tests")
+    variant("tests_c", default=False, description="Enable C tests")
```
Comment (albestro) [inline var/spack/repos/builtin/packages/elsi/package.py]: isn't better to use the `--test` flag, by removing these variants and enabling testing conditionally on `self.run_tests`?

Comment (RMeli) [reply var/spack/repos/builtin/packages/elsi/package.py]: Mabye I'm missing something, but these variants are to ensure the tests are built by CMake (which they are not by default).

Comment (albestro) [reply var/spack/repos/builtin/packages/elsi/package.py]: IMHO instead of using variants to enable them, you might use `self.run_tests` to enable them or not.

See some examples https://github.com/search?q=repo%3Aspack%2Fspack+self.run_tests&type=code

File: var/spack/repos/builtin/packages/elsi/package.py [hunk 2]
```diff
@@ -33,27 +33,58 @@ class Elsi(CMakePackage):
...
+    variant(
+        "internal_elpa_version",
+        default="2024",
+        values=("2024", "2023_11", "2023", "2021", "2020"),
+        description="Internal ELPA version",
+        multi=False,
+    )
```
Comment (msimberg) [inline var/spack/repos/builtin/packages/elsi/package.py:46]: Does this choose a bundled ELPA version? IMO it would be better to default the `use_external_x` variants to true, or even remove them and always use the external libraries. I know this is a breaking change, but using the internal bundled versions is a guaranteed road to hell. There's (controversial) precedent for removing the option to use bundled dependencies in https://github.com/spack/spack/pull/43210. This would also make it easier to model version constraints through spack's regular mechanisms instead of having to do it manually through the version variant.

I know this is a more intrusive change so I'm not expecting this to be changed now or ever, but as far as I could tell at least no other package currently requests explicitly any of the `~/+use_external_x` variants. This doesn't mean someone isn't using the variant in their environments of course...

**Review by msimberg — DISMISSED** [review]
Non-blocking comment and nothing in addition to @albestro's comment.

File: var/spack/repos/builtin/packages/elsi/package.py (hunk 2 above)
Comment (RMeli) [reply var/spack/repos/builtin/packages/elsi/package.py:46]: Yes, this variant is to choose the bundled ELPA version. The default is unfortunately still 2020.

I totally agree with your suggestion in principle, but ELPA is currently broken in Spack: https://github.com/spack/spack/issues/43902. Bundled libraries are non-optional in ELSI, so it is currently not possible to build cleanly with Spack using ELPA as an external dependency.

Comment (msimberg) [reply var/spack/repos/builtin/packages/elsi/package.py:46]: Yep, fair enough, whatever it takes to get it working at this point.

What do you mean by

> Bundled libraries are non-optional in ELSI

though? The options are clearly there, so do you mean the options are broken or something similar? I'm not sure I see the relation to https://github.com/spack/spack/issues/43902 (if there is a relation...)?

Comment (msimberg) [reply var/spack/repos/builtin/packages/elsi/package.py:46]: Yep, fair enough, whatever it takes to get it working at this point.

What do you mean by

> Bundled libraries are non-optional in ELSI

though? The options are clearly there, so do you mean the options are broken or something similar? I'm not sure I see the relation between bundled libraries and https://github.com/spack/spack/issues/43902 (if there is a relation...)?

Comment (RMeli) [reply var/spack/repos/builtin/packages/elsi/package.py:46]: Sorry, I was unclear. By non-optional I mean that you can't just say `elsi~elpa`. ELPA is bundled and has to be installed, either from the bundled versions or as an external package. So if you remove the possibility to use the bundled versions, external ELPA is always required.

Comment (msimberg) [reply var/spack/repos/builtin/packages/elsi/package.py:46]: Ok, that makes sense.

> So if you remove the possibility to use the bundled versions, external ELPA is always required.

This sounds like a win, but that you already know :sweat_smile: 

Looking at the ELSI CMake configuration, 2020 might be the default ELPA version, but it also looks like it only supports 2021 as an alternative version: https://github.com/ElectronicStructureLibrary/elsi-interface/blob/e27b7c925f30bc4486340e3cb6a32ffa4adc3de8/CMakeLists.txt#L103-L107. Or is there a branch somewhere that supports any version of ELPA when using the bundled version?

To summarize, my preference would be to completely remove the option of using bundled dependencies but I know this is drastic and I'm fine with not doing that. However, with the existing variants I would very much prefer to make the `use_external_x` variants default to false. This way users can also choose whatever version of ELPA they want (assuming it builds and is compatible...).

The above is also assuming I'm still not missing some weird corner case that requires preferring the bundled dependencies (e.g. ELSI not building with external ELPA).

Comment (RMeli) [reply var/spack/repos/builtin/packages/elsi/package.py:46]: > Or is there a branch somewhere that supports any version of ELPA when using the bundled version?

The official project is hosted on GitLab: https://gitlab.com/elsi_project I think the GitHub project is a mirror, and it has not been updated in 2 years.

> To summarize, my preference would be to completely remove the option of using bundled dependencies but I know this is drastic and I'm fine with not doing that. However, with the existing variants I would very much prefer to make the use_external_x variants default to false.

I agree, but to do this we need a working `elpa+cuda` in Spack.

I was hoping that who broke the package would fix it, but despite multiple pings from my part this is not the case. At this point it looks like I will have to fix it myself...

Comment (msimberg) [reply var/spack/repos/builtin/packages/elsi/package.py:46]: Thanks @RMeli for explaining this offline. The issue is that the external ELPA built through spack with CUDA does not build, whereas the internal bundled ELPA somehow does... So preferring the internal bundled ELPA indeed seems like a requirement for now (and may warrant a conflict?). Other than a possible addition of a conflict you can consider this thread resolved.

File: var/spack/repos/builtin/packages/elsi/package.py (hunk 1 above)
Comment (RMeli) [reply var/spack/repos/builtin/packages/elsi/package.py]: Thanks for the suggestion!

File: var/spack/repos/builtin/packages/elsi/package.py (hunk 2 above)
Comment (RMeli) [reply var/spack/repos/builtin/packages/elsi/package.py:46]: A glimps of hope from https://github.com/spack/spack/pull/44749, I'll look into using only external packages if they build well.

**Review by msimberg — DISMISSED** [review]

**Comment by RMeli** [comment]
I can't add a blocking review myself, but now that https://github.com/spack/spack/pull/44749 I'm looking into making ELPA an external dependency only as suggested by @msimberg.

**Comment by RMeli** [comment]
Let's get this one merged, then I'll test using external packages instead and make them the default behaviour as suggested.

**Review by rasolca — APPROVED** [review]
Looks fine.
Looking forward to PR to always use external ELPA...

File: var/spack/repos/builtin/packages/elsi/package.py [hunk 3]
```diff
@@ -33,27 +33,56 @@ class Elsi(CMakePackage):
...
+        if self.spec.satisfies("+use_external_ntpoly"):
+            libs_names.append("ntpoly")
+        if self.spec.satisfies("+use_external_superlu"):
+            libs_names.append("superlu-dist")
+
+        lib_paths, libs = [], []
+        for lib in libs_names:
+            lib_paths.extend(
+                [name.replace("-L", "") for name in self.spec[lib].libs.search_flags.split()]
+            )
```
Comment (albestro) [inline var/spack/repos/builtin/packages/elsi/package.py]:
```suggestion
            lib_paths.extend(self.spec[lib].libs.directories)
```

Does it make sense? Just for reference an example

```shell
ialberto@daint105:~> spack python
Spack version 0.23.0.dev0
Python 3.6.15, Linux x86_64
>>> from spack.spec import Spec
>>> s = Spec("dla-future")
>>> s.concretize()
==> Warning: The Cray platform, i.e. 'platform=cray', will be removed in Spack v0.23. All Cray machines will be then detected as 'platform=linux'.
>>> s.spec
Traceback (most recent call last):
  File "<console>", line 1, in <module>
AttributeError: 'Spec' object has no attribute 'spec'
>>> s['blas']
intel-oneapi-mkl@=2024.0.0%gcc@=11.2.0~cluster+envmods~ilp64+shared build_system=generic mpi_family=none threads=none arch=cray-cnl7-haswell ^[deptypes=build,link] berkeley-db@=18.1.40%gcc@=11.2.0+cxx~docs+stl build_system=autotools patches=26090f4,b231fcc arch=cray-cnl7-haswell ^[deptypes=build,link,run] bzip2@=1.0.8%gcc@=11.2.0~debug~pic+shared build_system=generic arch=cray-cnl7-haswell ^[deptypes=build] ca-certificates-mozilla@=2023-05-30%gcc@=11.2.0 build_system=generic arch=cray-cnl7-haswell ^[deptypes=build,run] cmake@=3.27.9%gcc@=11.2.0~doc+ncurses+ownlibs build_system=generic build_type=Release arch=cray-cnl7-haswell ^[deptypes=build,link] curl@=8.6.0%gcc@=11.2.0~gssapi~ldap~libidn2~librtmp~libssh~libssh2+nghttp2 build_system=autotools libs=shared,static tls=openssl arch=cray-cnl7-haswell ^[deptypes=build] diffutils@=3.10%gcc@=11.2.0 build_system=autotools arch=cray-cnl7-haswell ^[deptypes=build,run] findutils@=4.9.0%gcc@=11.2.0 build_system=autotools patches=440b954 arch=cray-cnl7-haswell ^[deptypes=link] gcc-runtime@=11.2.0%gcc@=11.2.0 build_system=generic arch=cray-cnl7-haswell ^[deptypes=build,link] gdbm@=1.23%gcc@=11.2.0 build_system=autotools arch=cray-cnl7-haswell ^[deptypes=build,link] gmake@=4.4.1%gcc@=11.2.0~guile build_system=generic arch=cray-cnl7-haswell ^[deptypes=build,link] hwloc@=2.9.1%gcc@=11.2.0~cairo~cuda~gl~libudev+libxml2~netloc~nvml~oneapi-level-zero~opencl+pci~rocm build_system=autotools libs=shared,static arch=cray-cnl7-haswell ^[deptypes=build,link virtuals=tbb] intel-tbb@=2021.9.0%gcc@=11.2.0~ipo+shared+tm build_system=cmake build_type=Release cxxstd=default generator=make patches=91755c6 arch=cray-cnl7-haswell ^[deptypes=build,link virtuals=iconv] libiconv@=1.17%gcc@=11.2.0 build_system=autotools libs=shared,static arch=cray-cnl7-haswell ^[deptypes=build,link] libpciaccess@=0.17%gcc@=11.2.0 build_system=autotools arch=cray-cnl7-haswell ^[deptypes=build,link] libsigsegv@=2.14%gcc@=11.2.0 build_system=autotools arch=cray-cnl7-haswell ^[deptypes=build] libtool@=2.4.7%gcc@=11.2.0 build_system=autotools arch=cray-cnl7-haswell ^[deptypes=build,link] libxml2@=2.10.3%gcc@=11.2.0+pic~python+shared build_system=autotools arch=cray-cnl7-haswell ^[deptypes=build,run] m4@=1.4.19%gcc@=11.2.0+sigsegv build_system=autotools patches=9dc5fbd,bfdffa7 arch=cray-cnl7-haswell ^[deptypes=build,link] ncurses@=6.4%gcc@=11.2.0~symlinks+termlib abi=none build_system=autotools patches=7a351bc arch=cray-cnl7-haswell ^[deptypes=build,link] nghttp2@=1.57.0%gcc@=11.2.0 build_system=autotools arch=cray-cnl7-haswell ^[deptypes=build,link] openssl@=3.2.1%gcc@=11.2.0~docs+shared build_system=generic certs=mozilla arch=cray-cnl7-haswell ^[deptypes=build,link,run] perl@=5.38.0%gcc@=11.2.0+cpanm+opcode+open+shared+threads build_system=generic patches=714e4d1 arch=cray-cnl7-haswell ^[deptypes=build virtuals=pkgconfig] pkg-config@=0.29.2%gcc@=11.2.0+internal_glib build_system=autotools arch=cray-cnl7-haswell ^[deptypes=build,link] readline@=8.2%gcc@=11.2.0 build_system=autotools patches=bbf97f1 arch=cray-cnl7-haswell ^[deptypes=build] util-macros@=1.19.3%gcc@=11.2.0 build_system=autotools arch=cray-cnl7-haswell ^[deptypes=build,link,run] xz@=5.4.6%gcc@=11.2.0~pic build_system=autotools libs=shared,static arch=cray-cnl7-haswell ^[deptypes=build,link virtuals=zlib-api] zlib-ng@=2.1.6%gcc@=11.2.0+compat+new_strategies+opt+pic+shared build_system=autotools arch=cray-cnl7-haswell
>>> s['blas'].libs
LibraryList(['/project/csstaff/ialberto/opt/spack-daint/opt/spack/cray-cnl7-haswell/gcc-11.2.0/intel-oneapi-mkl-2024.0.0-tlbhqpmmewnxlf6fhku3g2kfnjjf4uyc/mkl/2024.0/lib/libmkl_gf_lp64.so', '/project/csstaff/ialberto/opt/spack-daint/opt/spack/cray-cnl7-haswell/gcc-11.2.0/intel-oneapi-mkl-2024.0.0-tlbhqpmmewnxlf6fhku3g2kfnjjf4uyc/mkl/2024.0/lib/libmkl_sequential.so', '/project/csstaff/ialberto/opt/spack-daint/opt/spack/cray-cnl7-haswell/gcc-11.2.0/intel-oneapi-mkl-2024.0.0-tlbhqpmmewnxlf6fhku3g2kfnjjf4uyc/mkl/2024.0/lib/libmkl_core.so', '/usr/lib64/libpthread.so', '/usr/lib64/libm.so', '/usr/lib64/libdl.so'])
>>> s['blas'].libs.directories
['/project/csstaff/ialberto/opt/spack-daint/opt/spack/cray-cnl7-haswell/gcc-11.2.0/intel-oneapi-mkl-2024.0.0-tlbhqpmmewnxlf6fhku3g2kfnjjf4uyc/mkl/2024.0/lib', '/usr/lib64']
>>> s['blas'].libs.libraries
['/project/csstaff/ialberto/opt/spack-daint/opt/spack/cray-cnl7-haswell/gcc-11.2.0/intel-oneapi-mkl-2024.0.0-tlbhqpmmewnxlf6fhku3g2kfnjjf4uyc/mkl/2024.0/lib/libmkl_gf_lp64.so', '/project/csstaff/ialberto/opt/spack-daint/opt/spack/cray-cnl7-haswell/gcc-11.2.0/intel-oneapi-mkl-2024.0.0-tlbhqpmmewnxlf6fhku3g2kfnjjf4uyc/mkl/2024.0/lib/libmkl_sequential.so', '/project/csstaff/ialberto/opt/spack-daint/opt/spack/cray-cnl7-haswell/gcc-11.2.0/intel-oneapi-mkl-2024.0.0-tlbhqpmmewnxlf6fhku3g2kfnjjf4uyc/mkl/2024.0/lib/libmkl_core.so', '/usr/lib64/libpthread.so', '/usr/lib64/libm.so', '/usr/lib64/libdl.so']
>>> s['blas'].libs.search_flags
'-L/project/csstaff/ialberto/opt/spack-daint/opt/spack/cray-cnl7-haswell/gcc-11.2.0/intel-oneapi-mkl-2024.0.0-tlbhqpmmewnxlf6fhku3g2kfnjjf4uyc/mkl/2024.0/lib -L/usr/lib64'
>>> s['blas'].libs.directories
['/project/csstaff/ialberto/opt/spack-daint/opt/spack/cray-cnl7-haswell/gcc-11.2.0/intel-oneapi-mkl-2024.0.0-tlbhqpmmewnxlf6fhku3g2kfnjjf4uyc/mkl/2024.0/lib', '/usr/lib64']
```

and some API reference

https://spack.readthedocs.io/en/latest/llnl.util.html#llnl.util.filesystem.FileList
https://spack.readthedocs.io/en/latest/llnl.util.html#llnl.util.filesystem.LibraryList

File: var/spack/repos/builtin/packages/elsi/package.py [hunk 4]
```diff
@@ -33,27 +33,56 @@ class Elsi(CMakePackage):
...
+            libs.extend([os.path.basename(name) for name in self.spec[lib].libs.libraries])
```
Comment (albestro) [inline var/spack/repos/builtin/packages/elsi/package.py]: You can give a try to `spec['lib'].libs.names`

File: var/spack/repos/builtin/packages/elsi/package.py (hunk 3 above)
Comment (RMeli) [reply var/spack/repos/builtin/packages/elsi/package.py]: Thank you, much nicer!

File: var/spack/repos/builtin/packages/elsi/package.py [hunk 5]
```diff
@@ -33,27 +33,56 @@ class Elsi(CMakePackage):
     variant("enable_sips", default=False, description="Enable SLEPc-SIPs support")
     variant("use_external_elpa", default=False, description="Build ELPA using SPACK")
     variant("use_external_ntpoly", default=False, description="Build NTPoly using SPACK")
-    variant("use_external_omm", default=False, description="Use external libOMM and MatrixSwitch")
```
Comment (RMeli) [inline var/spack/repos/builtin/packages/elsi/package.py:36]: `libOMM` is not available in Spack. It can be found here: https://gitlab.com/ElectronicStructureLibrary/omm/libomm

This will be tackled (eventually) in another PR.

**Comment by RMeli** [comment]
In the end, the following builds and passes all tests:
```
==> Concretized 1 spec:
 -   mxtxjex  elsi@master%gcc@11.4.0 fflags=-fallow-argument-mismatch +add_underscore~cuda~enable_pexsi~enable_sips~ipo+use_external_elpa+use_external_ntpoly+use_external_superlu+use_mpi_iallgather build_system=cmake build_type=Release elpa2_kernel=none generator=make internal_elpa_version=2024 arch=linux-ubuntu22.04-broadwell
[+]  m2ej5ad      ^cmake@3.27.9%gcc@11.4.0~doc+ncurses+ownlibs build_system=generic build_type=Release arch=linux-ubuntu22.04-broadwell
[+]  4werial          ^curl@8.7.1%gcc@11.4.0~gssapi~ldap~libidn2~librtmp~libssh~libssh2+nghttp2 build_system=autotools libs=shared,static tls=openssl arch=linux-ubuntu22.04-broadwell
[+]  icwsrao              ^nghttp2@1.57.0%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-broadwell
[+]  f6jafzv                  ^diffutils@3.10%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-broadwell
[+]  tp5ttwn              ^openssl@3.3.0%gcc@11.4.0~docs+shared build_system=generic certs=mozilla arch=linux-ubuntu22.04-broadwell
[+]  pnlaiqe                  ^ca-certificates-mozilla@2023-05-30%gcc@11.4.0 build_system=generic arch=linux-ubuntu22.04-broadwell
[+]  buntu7p          ^ncurses@6.5%gcc@11.4.0~symlinks+termlib abi=none build_system=autotools patches=7a351bc arch=linux-ubuntu22.04-broadwell
[+]  bjkgkde          ^zlib-ng@2.1.6%gcc@11.4.0+compat+new_strategies+opt+pic+shared build_system=autotools arch=linux-ubuntu22.04-broadwell
[+]  bqpu4fs      ^elpa@2024.03.001%gcc@11.4.0~autotune~cuda+mpi~openmp~rocm build_system=autotools arch=linux-ubuntu22.04-broadwell
[+]  wccsgve          ^libtool@2.4.7%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-broadwell
[+]  4fyn4it              ^m4@1.4.19%gcc@11.4.0+sigsegv build_system=autotools patches=9dc5fbd,bfdffa7 arch=linux-ubuntu22.04-broadwell
[+]  b56l5px                  ^libsigsegv@2.14%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-broadwell
[+]  gp2ha5u          ^python@3.11.7%gcc@11.4.0+bz2+crypt+ctypes+dbm~debug+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tkinter+uuid+zlib build_system=generic patches=13fa8bf,b0615b2,ebdca64,f2fd060 arch=linux-ubuntu22.04-broadwell
[+]  6a524bo              ^bzip2@1.0.8%gcc@11.4.0~debug~pic+shared build_system=generic arch=linux-ubuntu22.04-broadwell
[+]  ktdyesm              ^expat@2.6.2%gcc@11.4.0+libbsd build_system=autotools arch=linux-ubuntu22.04-broadwell
[+]  ea5pkue                  ^libbsd@0.12.1%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-broadwell
[+]  vxnqbqk                      ^libmd@1.0.4%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-broadwell
[+]  noq3ou6              ^gdbm@1.23%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-broadwell
[+]  4dp776w              ^gettext@0.22.5%gcc@11.4.0+bzip2+curses+git~libunistring+libxml2+pic+shared+tar+xz build_system=autotools arch=linux-ubuntu22.04-broadwell
[+]  2edmqor                  ^tar@1.34%gcc@11.4.0 build_system=autotools zip=pigz arch=linux-ubuntu22.04-broadwell
[+]  svawaft                      ^pigz@2.8%gcc@11.4.0 build_system=makefile arch=linux-ubuntu22.04-broadwell
[+]  5oorsex                      ^zstd@1.5.6%gcc@11.4.0+programs build_system=makefile compression=none libs=shared,static arch=linux-ubuntu22.04-broadwell
[+]  cecogma              ^libffi@3.4.6%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-broadwell
[+]  jsryzyv              ^libxcrypt@4.4.35%gcc@11.4.0~obsolete_api build_system=autotools patches=4885da3 arch=linux-ubuntu22.04-broadwell
[+]  t57e3hu              ^readline@8.2%gcc@11.4.0 build_system=autotools patches=bbf97f1 arch=linux-ubuntu22.04-broadwell
[+]  sdp3aq6              ^sqlite@3.43.2%gcc@11.4.0+column_metadata+dynamic_extensions+fts~functions+rtree build_system=autotools arch=linux-ubuntu22.04-broadwell
[+]  npukef7              ^util-linux-uuid@2.38.1%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-broadwell
[+]  vhuojrm              ^xz@5.4.6%gcc@11.4.0~pic build_system=autotools libs=shared,static arch=linux-ubuntu22.04-broadwell
[+]  7pmsvsm      ^gcc-runtime@11.4.0%gcc@11.4.0 build_system=generic arch=linux-ubuntu22.04-broadwell
[e]  a6gbmc4      ^glibc@2.35%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-broadwell
[+]  kzsslwd      ^gmake@4.4.1%gcc@11.4.0~guile build_system=generic arch=linux-ubuntu22.04-broadwell
[+]  ax65zzy      ^mpich@4.2.1%gcc@11.4.0~argobots~cuda+fortran+hwloc+hydra+libxml2+pci~rocm+romio~slurm~verbs+wrapperrpath~xpmem build_system=autotools datatype-engine=auto device=ch3 netmod=tcp pmi=default arch=linux-ubuntu22.04-broadwell
[+]  j7hgdgb          ^findutils@4.9.0%gcc@11.4.0 build_system=autotools patches=440b954 arch=linux-ubuntu22.04-broadwell
[+]  ni4y45c          ^hwloc@2.9.1%gcc@11.4.0~cairo~cuda~gl~libudev+libxml2~netloc~nvml~oneapi-level-zero~opencl+pci~rocm build_system=autotools libs=shared,static arch=linux-ubuntu22.04-broadwell
[+]  7bhosfx          ^libpciaccess@0.17%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-broadwell
[+]  vq2625p              ^util-macros@1.19.3%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-broadwell
[+]  32ynd5v          ^libxml2@2.10.3%gcc@11.4.0+pic~python+shared build_system=autotools arch=linux-ubuntu22.04-broadwell
[+]  pur6zt3          ^pkgconf@2.2.0%gcc@11.4.0 build_system=autotools arch=linux-ubuntu22.04-broadwell
[+]  vaik6qz      ^netlib-scalapack@2.2.0%gcc@11.4.0~ipo~pic+shared build_system=cmake build_type=Release generator=make patches=072b006,1c9ce5f,244a9aa arch=linux-ubuntu22.04-broadwell
[+]  sscc2cp      ^ntpoly@3.1.0%gcc@11.4.0~ipo+shared build_system=cmake build_type=Release generator=make arch=linux-ubuntu22.04-broadwell
[+]  7b5sulo      ^openblas@0.3.26%gcc@11.4.0~bignuma~consistent_fpcsr+dynamic_dispatch+fortran~ilp64+locking+pic+shared build_system=makefile symbol_suffix=none threads=none arch=linux-ubuntu22.04-broadwell
[+]  li6qitn          ^perl@5.38.0%gcc@11.4.0+cpanm+opcode+open+shared+threads build_system=generic patches=714e4d1 arch=linux-ubuntu22.04-broadwell
[+]  w4qt5lz              ^berkeley-db@18.1.40%gcc@11.4.0+cxx~docs+stl build_system=autotools patches=26090f4,b231fcc arch=linux-ubuntu22.04-broadwell
[+]  6et3dzr      ^superlu-dist@8.2.1%gcc@11.4.0~cuda~int64~ipo~openmp+parmetis~rocm+shared build_system=cmake build_type=Release generator=make arch=linux-ubuntu22.04-broadwell
[+]  f4hipts          ^metis@5.1.0%gcc@11.4.0~gdb~int64~ipo~real64+shared build_system=cmake build_type=Release generator=make patches=4991da9,93a7903,b1225da arch=linux-ubuntu22.04-broadwell
[+]  62kqq4u          ^parmetis@4.0.3%gcc@11.4.0~gdb~int64~ipo+shared build_system=cmake build_type=Release generator=make patches=4f89253,50ed208,704b84f arch=linux-ubuntu22.04-broadwell
```

I haven't tried other versions of the external dependencies, but given that at least one configuration works I switched the default to use external dependencies.

**Review by albestro — DISMISSED** [review]

**Review by msimberg — DISMISSED** [review]

**Comment by msimberg** [comment]
@RMeli style checks are unhappy, but otherwise this can be merged.

**Comment by albestro** [comment]
@spackbot fix style

**Comment by RMeli** [comment]
@spackbot fix style

**Comment by RMeli** [comment]
@spackbot hello!

**Review by alalazo — APPROVED** [review]
